In [ ]:
# This mounts your Google Drive to the Colab VM.
from google.colab import drive
drive.mount('/content/drive')

# TODO: Enter the foldername in your Drive where you have saved the unzipped
# assignment folder, e.g. 'cs231n/assignments/assignment2/'
FOLDERNAME = "cs224u/cs224u/"
assert FOLDERNAME is not None, "[!] Enter the foldername."

# Now that we've mounted your Drive, this ensures that
# the Python interpreter of the Colab VM can load
# python files from within it.
import sys
sys.path.append('/content/drive/My Drive/{}'.format(FOLDERNAME))

# # This downloads the CIFAR-10 dataset to your Drive
# # if it doesn't already exist.
# %cd /content/drive/My\ Drive/$FOLDERNAME/cs231n/datasets/
# !bash get_datasets.sh
# %cd /content/drive/My\ Drive/$FOLDERNAME

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.1/7.1 MB 58.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 236.8/236.8 kB 31.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.8/7.8 MB 108.0 MB/s eta 0:00:00


In [ ]:
from transformers import pipeline
MODEL = "jy46604790/Fake-News-Bert-Detect"
clf = pipeline("text-classification", model=MODEL, tokenizer=MODEL)


Xformers is not installed correctly. If you want to use memorry_efficient_attention to accelerate training use the following command to install Xformers
pip install xformers.


In [ ]:
clf

In [ ]:
text = "Indonesian police have recaptured a U.S. citizen who escaped a week ago from an overcrowded prison on the holiday island of Bali, the jail s second breakout of foreign inmates this year.  Cristian Beasley from California was rearrested on Sunday, Badung Police chief Yudith Satria Hananta said, without providing further details.  Beasley was a suspect in crimes related to narcotics but had not been sentenced when he escaped from Kerobokan prison in Bali last week. The 32-year-old is believed to have cut through bars in the ceiling of his cell before scaling a perimeter wall of the prison in an area being refurbished. The Kerobokan prison, about 10 km (six miles) from the main tourist beaches in the Kuta area, often holds foreigners facing drug-related charges. Representatives of Beasley could not immediately be reached for comment. In June, an Australian, a Bulgarian, an Indian and a Malaysian tunneled to freedom about 12 meters (13 yards) under Kerobokan prison s walls. The Indian and the Bulgarian were caught soon after in neighboring East Timor, but Australian Shaun Edward Davidson and Malaysian Tee Kok King remain at large. Davidson has taunted authorities by saying he was enjoying life in various parts of the world, in purported posts on Facebook.  Kerobokan has housed a number of well-known foreign drug convicts, including Australian Schappelle Corby, whose 12-1/2-year sentence for marijuana smuggling got huge media attention."



In [ ]:
result = clf(text)
result


[{'label': 'LABEL_1', 'score': 0.9994995594024658}]

In [ ]:
import pandas as pd


In [ ]:
df_fake = pd.read_csv("/content/drive/MyDrive/cs244u/cs224u/Fake.csv")
df_fake.head()

title  \
0   Donald Trump Sends Out Embarrassing New Year’...   
1   Drunk Bragging Trump Staffer Started Russian ...   
2   Sheriff David Clarke Becomes An Internet Joke...   
3   Trump Is So Obsessed He Even Has Obama’s Name...   
4   Pope Francis Just Called Out Donald Trump Dur...   

                                                text subject  \
0  Donald Trump just couldn t wish all Americans ...    News   
1  House Intelligence Committee Chairman Devin Nu...    News   
2  On Friday, it was revealed that former Milwauk...    News   
3  On Christmas day, Donald Trump announced that ...    News   
4  Pope Francis used his annual Christmas Day mes...    News   

                date  
0  December 31, 2017  
1  December 31, 2017  
2  December 30, 2017  
3  December 29, 2017  
4  December 25, 2017

In [ ]:
df_true = pd.read_csv("/content/drive/MyDrive/cs244u/cs224u/True.csv")
df_true.head()

title  \
0  As U.S. budget fight looms, Republicans flip t...   
1  U.S. military to accept transgender recruits o...   
2  Senior U.S. Republican senator: 'Let Mr. Muell...   
3  FBI Russia probe helped by Australian diplomat...   
4  Trump wants Postal Service to charge 'much mor...   

                                                text       subject  \
0  WASHINGTON (Reuters) - The head of a conservat...  politicsNews   
1  WASHINGTON (Reuters) - Transgender people will...  politicsNews   
2  WASHINGTON (Reuters) - The special counsel inv...  politicsNews   
3  WASHINGTON (Reuters) - Trump campaign adviser ...  politicsNews   
4  SEATTLE/WASHINGTON (Reuters) - President Donal...  politicsNews   

                 date  
0  December 31, 2017   
1  December 29, 2017   
2  December 31, 2017   
3  December 30, 2017   
4  December 29, 2017

In [ ]:
#Label 0 = fake, Label 1 = true
ex1 = (df_fake.iloc[1]["text"])
#split by 500 and join
ex1_tokens = ex1.split(" ")
ex1 = " ".join(ex1_tokens[:200])
result = clf(ex1)
result


[{'label': 'LABEL_0', 'score': 0.9997329115867615}]

In [ ]:
#Label 0 = fake, Label 1 = true
ex2 = (df_true.iloc[1]["text"])
#split by 500 and join
ex2_tokens = ex2.split(" ")
ex2 = " ".join(ex2_tokens[:200])
result = clf(ex2)
result

[{'label': 'LABEL_1', 'score': 0.9999136924743652}]

In [ ]:
result = clf([ex1, ex2])
print(result)

[{'label': 'LABEL_0', 'score': 0.9997329115867615}, {'label': 'LABEL_1', 'score': 0.9999136924743652}]


In [ ]:
df_fake.shape
NUM_FAKE, _ = df_fake.shape

In [ ]:
NUM_TRUE, _ = df_true.shape

In [ ]:
#half of each of these dataframes
first_half_fake = df_fake.iloc[:120]
first_half_fake.shape
first_half_true = df_true.iloc[:120]
first_half_true.shape

(120, 4)

In [ ]:
def get_first_500(df_of_data):
  data = []
  for index, row in df_of_data.iterrows():
    article = df_of_data.iloc[index]["text"]
    article_tokens = article.split(" ")
    article = " ".join(article_tokens[:200])
    data.append(article)
  return data
fake_data = get_first_500(first_half_fake)
true_data = get_first_500(first_half_true)



In [ ]:
len(fake_data)

23481

In [ ]:
def evaluate(data):
  #calculate the total number of Label 0 and Label 1
  label_0 = 0
  label_1 = 0
  a =0
  for article in data:
    print(a)
    result = clf(article)
    if result[0]["label"] == 'LABEL_0':
      label_0 += 1
    elif result[0]["label"] == 'LABEL_1':
      label_1 += 1
    a += 1
  return label_0, label_1

In [ ]:
fake_found, true_found = evaluate(fake_data)
print(fake_found, true_found,NUM_FAKE )

In [ ]:
fake_found, true_found = evaluate(true_data)
print(fake_found, true_found,NUM_FAKE) 

120 0 23481


In [ ]:
#CAN WE JUST ADD PROMPT BFORE PASSIN IN KAGGLE DATA? DSP

# Evaluating model on datasets

In [ ]:
'''
This cell brings in the model for global use
'''
from transformers import pipeline
MODEL = "jy46604790/Fake-News-Bert-Detect"
fact_checker = pipeline("text-classification", model=MODEL, tokenizer=MODEL)


In [ ]:
'''
This method takes in a csv pathname through its parameter
and takes a batch of data from it. 
This is because these datasets are too large
to appropriately run and evaluate.
'''

def take_sample_of_dataset(dataset_csv_path):
  df = pd.read_csv(dataset_csv_path)
  batch = df.iloc[:120]
  return batch

In [ ]:
'''
This method takes in a dataframe through its parameters
and a boolean for whether this is the true or fake news dataset. 
It returns the number of actual fake or true instances
that we should expect i.e. gold label
'''
def configure_true_fake_metrics(dataframe, is_fake):
  num_expected_fake = 0
  num_expected_true = 0
  if is_fake:
    num_expected_fake = dataframe.shape[0]
  else:
    num_expected_true = dataframe.shape[0]
  
  return num_expected_true, num_expected_fake


In [ ]:
'''
Since the dataset is quite large, and each text
entry is large, we take only the first 200 
tokens of each entry. 
This method extracts the first 200 tokens
and constructs a list of article entries.
'''
def get_first_500(df_of_data):
  data = []
  for index, row in df_of_data.iterrows():
    article = df_of_data.iloc[index]["text"]
    article_tokens = article.split(" ")
    article = " ".join(article_tokens[:200])
    data.append(article)
  return data

In [ ]:
'''
This method counts the number of times the model categorizes
an article as 'fake' or 'true' give by the respective labels
'Label_0' and 'Label_1'. 
'''
def evaluate(dataset_path, is_fake):
  #calculate the total number of Label 0 and Label 1
  label_0 = 0
  label_1 = 0
  a =0
  #bring in dataset
  df = take_sample_of_dataset(dataset_path)
  #get ground truth values
  actual_true, actual_fake = configure_true_fake_metrics(df, is_fake)
  #convert data to ingestible format
  data = get_first_500(df)

  for article in data:
    if a % 30 == 0:
      print("In cycle a = ", a)
    result = fact_checker(article)
    if result[0]["label"] == 'LABEL_0':
      label_0 += 1
    elif result[0]["label"] == 'LABEL_1':
      label_1 += 1
    a += 1
  return label_0, label_1, actual_true, actual_fake

In [ ]:
'''
This cell conducts evaluation of the model 
on several datasets to examine performance
'''
found_fake, found_true, actual_true, actual_fake = evaluate("/content/drive/MyDrive/cs244u/cs224u/Fake.csv", is_fake=True)
print("Found fake instances = " , found_fake, "Found true instances = ", found_true, "Actual true instances = ",actual_true, found_fake, "Actual fake instances = ", actual_fake)

found_fake, found_true, actual_true, actual_fake = evaluate("/content/drive/MyDrive/cs244u/cs224u/True.csv", is_fake=False)
print("Found fake instances = " , found_fake, "Found true instances = ", found_true, "Actual true instances = ",actual_true, found_fake, "Actual fake instances = ", actual_fake)

In cycle a =  0
In cycle a =  30
In cycle a =  60
In cycle a =  90
Found fake instances =  120 Found true instances =  0 120 Actual true instances =  0 120 Actual fake instances =  120
In cycle a =  0
In cycle a =  30
In cycle a =  60
In cycle a =  90
Found fake instances =  0 Found true instances =  120 0 Actual true instances =  120 0 Actual fake instances =  0


In [ ]:
found_fake, found_true, actual_true, actual_fake = evaluate("/content/drive/MyDrive/cs244u/cs224u/DataSet_Misinfo_TRUE.csv", is_fake=False)
print("Found fake instances = " , found_fake, "Found true instances = ", found_true, "Actual true instances = ",actual_true, "Actual fake instances = ", actual_fake)


In cycle a =  0
In cycle a =  30
In cycle a =  60
In cycle a =  90
Found fake instances =  9 Found true instances =  111 Actual true instances =  120 Actual fake instances =  0


For the Dataset_Misinfo_True dataset, the model found this: 
Found fake instances =  9 Found true instances =  111 Actual true instances =  120 Actual fake instances =  0
There are actually 120 true instances!


```
# This is formatted as code
```



In [ ]:
found_fake, found_true, actual_true, actual_fake = evaluate("/content/drive/MyDrive/cs244u/cs224u/DataSet_Misinfo_FAKE.csv", is_fake=True)
print("Found fake instances = " , found_fake, "Found true instances = ", found_true, "Actual true instances = ",actual_true, "Actual fake instances = ", actual_fake)


In cycle a =  0
In cycle a =  30
In cycle a =  60
In cycle a =  90
Found fake instances =  120 Found true instances =  0 Actual true instances =  0 Actual fake instances =  120


For the Dataset_Misinfo_True dataset, the model found this: 
Found fake instances =  120 Found true instances =  0 Actual true instances =  0 Actual fake instances =  120


# Keyword Identification for Event Categorization with Demonstrated-Search-Predict, GPT-3.5Turbo

In [ ]:
try: 
    # This library is our indicator that the required installs
    # need to be done.
    import datasets
    root_path = '.'
except ModuleNotFoundError:
    !git clone https://github.com/cgpotts/cs224u/
    !pip install -r cs224u/requirements.txt
    root_path = 'dsp'

Cloning into 'cs224u'...
remote: Enumerating objects: 2216, done.
remote: Counting objects: 100% (124/124), done.
remote: Compressing objects: 100% (78/78), done.
remote: Total 2216 (delta 55), reused 83 (delta 46), pack-reused 2092
Receiving objects: 100% (2216/2216), 41.49 MiB | 24.50 MiB/s, done.
Resolving deltas: 100% (1354/1354), done.
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Cloning https://github.com/stanfordnlp/dsp to /tmp/pip-req-build-18j2jiug
  Running command git clone --filter=blob:none --quiet https://github.com/stanfordnlp/dsp /tmp/pip-req-build-18j2jiug
  Resolved https://github.com/stanfordnlp/dsp to commit 298fac8937fdf7326d1b85546ecd41f099ed33e2
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 887.5/887.5 MB 1.1 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.2/24.2 MB 72.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.3/6.3 MB 48.2 MB

In [ ]:
import cohere
from datasets import load_dataset
import openai
import os
import dsp
import numpy as np

In [ ]:
import locale
def getpreferredencoding(do_setlocale = True):
    return "UTF-8"
locale.getpreferredencoding = getpreferredencoding

In [ ]:
!pip install faiss-gpu
!pip install faiss-cpu


Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 85.5/85.5 MB 20.2 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.6/17.6 MB 73.1 MB/s eta 0:00:00


In [ ]:
os.environ["DSP_NOTEBOOK_CACHEDIR"] = os.path.join(root_path, 'cache')

# openai_key = os.getenv('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')  # or replace with your API key (optional)
openai_key = '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'  # or replace with your API key (optional)
openai.api_key="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"
cohere_key = 'oPdNmragJYgt484JGwQVh2wQEHtbxoK1g4Dh0lBb'  # or replace with your API key (optional)

colbert_server = 'http://ec2-44-228-128-229.us-west-2.compute.amazonaws.com:8893/api/search'

In [ ]:
lm = dsp.GPT3(model='text-davinci-002', api_key=openai_key)
rm = dsp.ColBERTv2(url=colbert_server)

dsp.settings.configure(lm=lm, rm=rm)

In [ ]:
df = take_sample_of_dataset("/content/drive/MyDrive/cs244u/cs224u/Fake.csv")
#convert data to ingestible format
data = get_first_500(df)

train_dataset1 =   data

In [ ]:
[d["root"] for d in openai.Model.list()["data"]]

In [ ]:
#Prepare dataset 
election_keywords = ["Donald Trump", "Trump", "Biden", "Joe", "Joe Biden", "election", "White House", "government", "congress", "Congress", "Nancy Pelosi", "Pelosi",]


In [ ]:
train_dataset1 = [dsp.Example(question="Can you categorize this information into one of the following categories: Elections or Political Information, Celebrity Gossip, Conspiracy Theories About Major Events, Misrepresenting medical information", context=x) for x in train_dataset1]
print(dsp.sample(train_dataset1, k=7))

[{'question': 'Can you categorize this information into one of the following categories: Elections or Political Information, Celebrity Gossip, Conspiracy Theories About Major Events, Misrepresenting medical information', 'context': {'question': 'Can you categorize this information into one of the following categories: Elections or Political Information, Celebrity Gossip, Conspiracy Theories About Major Events, Misrepresenting medical information', 'context': 'Donald Trump thinks if he rage-tweets about NFL players who choose to kneel instead of stand during the national anthem as a form of silent protest against racial injustice and police brutality, it will all go away but by doing that, he inadvertently highlights that there is a problem. The former reality show star dropped a racial slur yesterday in front of Native American heroes, then today, he jumped on Twitter to yell at black people. We see a pattern here. At least 24 players kneeling this weekend at NFL stadiums that are now 

In [ ]:
Question = dsp.Type(prefix="Question:", desc="${the question to be answered}")
Answer = dsp.Type(prefix="Answer:", desc="${a short factoid answer, often between 1 and 5 words}", format=dsp.format_answers)
Context = dsp.Type(prefix="Context:\n",desc="${sources that may contain relevant content}",format=dsp.passages2text)
qa_template = dsp.Template(instructions="Answer questions with short factoid answers.", question=Question(), answer=Answer())

In [ ]:
def fs_openqa_w_context(question: str, context: str) -> str:
  demos = dsp.sample(train_dataset1, k=7)
  example = dsp.Example(question=question, demos=demos, context=context)
  example, completions = dsp.generate(qa_template, temperature=0.2)(example, stage='qa')
  return completions.answer

In [ ]:
gens = []
for question in train_dataset1:
  category = fs_openqa_w_context(train_dataset1[8].question, train_dataset1[8].context)
  
print(train_dataset1[8].context)
print()

{'question': 'Can you categorize this information into one of the following categories: Elections or Political Information, Celebrity Gossip, Conspiracy Theories About Major Events, Misrepresenting medical information', 'context': 'Many people have raised the alarm regarding the fact that Donald Trump is dangerously close to becoming an autocrat. The thing is, democracies become autocracies right under the people s noses, because they can often look like democracies in the beginning phases. This was explained by Republican David Frum just a couple of months into Donald Trump s presidency, in a piece in The Atlantic called  How to Build an Autocracy. In fact, if you really look critically at what is happening right now   the systematic discrediting of vital institutions such as the free press and the Federal Bureau of Investigation as well the direct weaponization of the Department of Justice in order to go after Trump s former political opponent, 2016 Democratic nominee Hillary Clinton

#Format data to be sent back


In [ ]:
return_message = {}
incoming_string = "Donald Trump was a president of the U.S."
result = fact_checker(incoming_string)
inc_ex = dsp.Example(question="Can you categorize this information into one of the following categories: Elections or Political Information, Celebrity Gossip, Conspiracy Theories About Major Events, Misrepresenting medical information", context=incoming_string)
category = fs_openqa_w_context(inc_ex.question, inc_ex.context)


if result[0]["label"] == 'LABEL_0':
  finding = {"label": "fake", "evidence" : incoming_string}
elif result[0]["label"] == 'LABEL_1':
  finding = {"label": "true", "evidence" : incoming_string}
return_message[category] = finding

print(return_message)


{'Elections or Political Information': {'label': 'true', 'evidence': 'Donald Trump was a president of the U.S.'}}
